filebeat/input/journald/input.go

```diff
@@ -204,3 +218,45 @@ func seekBy(log *logp.Logger, cp checkpoint, seek, defaultSeek journalread.SeekM
 	}
 	return mode, cp.Position
 }
+
+// readerAdapter is an adapter so journalread.Reader can
+// behave like reader.Reader
+type readerAdapter struct {
+	r        *journalread.Reader
+	canceler input.Canceler
+}
+
+func (r *readerAdapter) Close() error {
+	return r.r.Close()
+}
+
+func (r *readerAdapter) Next() (reader.Message, error) {
```

Comment:
`Meta` is a legacy attribute we would like to get rid of. So please avoid using it.

Instead of putting partial state information to `Meta`, you should add a `Private` attribute to `Message` and save the state/checkpoint information there in this `Next` function.

Example

```golang

func (r *readerAdapter) Next() (reader.Message, error) {
   // ...
    m := reader.Message{
        Private: checkpoint{
            Version: cursorVersion,
            RealtimeTimestamp: data.RealtimeTimestamp,
            MonotonicTimestamp: data.MonotonicTimestamp,
            Position: data.Cursor,
            Path: r.path,
      }
    }
```

Then it can be published like this:
```golang
if err := publisher.Publish(event, event.Private); err != nil {
```

Reply:
Done, I force pushed the changes alongside other improvements. I hope I got it right.